### PR TITLE
Add swaps savings calculation

### PR DESCRIPTION
**Pull request by rekmarks**
This PR adds savings calculations for swaps, in the swaps controller.

The function previously named `_findTopQuoteAggId` has been renamed `_findTopQuoteAndCalculateSavings`. This function already made the calculations to find the best quote by overall ETH value.

With very little added work, we now also calculate the overall swap fees in ETH, which include gas costs and all exchange fees except our own, which proportionally identical across all quotes and already subtracted by the time the quotes are received.

Finally, we calculate the savings for the best quote. In the UI, we will display these savings along with our fee. We calculate savings relative to the median fee and the median trade value for all quotes, compared to that of the best quote.

Here are the equations in pseudocode, assuming all units converted to ETH:

```
total_eth_cost = swap_tx_gas_cost + approval_tx_gas_cost + swap_tx_eth_sent
```

We say `total_eth_cost`, because it does not include the value of the source token unless the source token is ETH.

`swap_tx_eth_sent` includes third-party fees and, if the source token is ETH, the amount of ETH swapped.

```
trade_fee = total_eth_cost - (source_token_is_eth ? amount_of_eth_swapped : 0)

trade_value = destination_token_eth_value - total_eth_cost
```

For each quote, we store `trade_fee` and `trade_value` for use in subsequent calculations.

We want the trade value to be as _high_ as possible, so we subtract the median from the best value.
```
performance_savings = best_quote_trade_value - median(all_trade_values)
```

We want the trade fees to be as _low_ as possible, so we subtract the fees of the best quote from the median fee of all quotes.
```
fee_savings = median(all_trade_fees) - best_quote_trade_fee
```

Note that either component of total_savings can be negative, and we often observe one negative and one positive in practice.
```
total_savings = performance_savings + fee_savings
```

### CODE REVIEW

File: app/scripts/controllers/swaps.js [hunk 1]
```diff
@@ -455,22 +485,50 @@ export default class SwapsController {
...
+      // Where the average is the statistical mean:
+      //   sumOfAllValues / numberOfValues
+
+      savings.performance = ethValueOfTradeForBestQuote.minus(
+        sumOfAllTradeValues.dividedBy(numQuotes),
+        10,
+      )
+
+      savings.fees = ethFeeForBestQuote.minus(
+        sumOfAllEthFees.dividedBy(numQuotes),
```
Comment (danjm) [inline app/scripts/controllers/swaps.js:522]: This should actually be:
```
savings.fees = sumOfAllEthFees.dividedBy(numQuotes).minus(
    ethFeeForBestQuote,
    10,
)
```

Comment (danjm) [reply app/scripts/controllers/swaps.js:522]: If I would have spent $10 on an apple from Alice, and $15 on an apple from Bob, but I ended up buy an apple from Carol for $5, then the average cost of apples is $10 and my savings against the average is $5.

(($10 + $15 + $5) / 3) - $5 = $5

Comment (rekmarks) [reply app/scripts/controllers/swaps.js]: This has been resolved!

File: app/scripts/controllers/swaps.js [hunk 2]
```diff
@@ -394,59 +395,88 @@ export default class SwapsController {
...
+            toDenomination: 'ETH',
+            fromNumericBase: 'BN',
+            numberOfDecimals: 6,
+          },
+        )
+        : totalEthCost
+
       const tokenConversionRate = tokenConversionRates[destinationToken]
       const ethValueOfTrade =
         destinationTokenInfo.symbol === 'ETH'
```
Comment (danjm) [inline app/scripts/controllers/swaps.js:478]: Now that we are doing the `sourceToken === ETH_SWAPS_TOKEN_ADDRESS` comparison above, maybe we should make the comparison here consistent and do `destinationToken === ETH_SWAPS_TOKEN_ADDRESS`

File: app/scripts/controllers/swaps.js [hunk 3]
```diff
@@ -394,59 +395,88 @@ export default class SwapsController {
...
-          ? calcTokenAmount(destinationAmount, 18).minus(ethFee, 10)
+          ? calcTokenAmount(destinationAmount, 18).minus(totalEthCost, 10)
```
Comment (danjm) [inline app/scripts/controllers/swaps.js:479]: Review note: this is functionally equivalent. What used to be called `ethFee` is now `totalEthCost`. (What is now `ethFee` is now different from `totalEthCost` when `sourceToken === ETH_SWAPS_TOKEN_ADDRESS`)

File: app/scripts/controllers/swaps.js [hunk 4]
```diff
@@ -455,22 +485,51 @@ export default class SwapsController {
               ),
               10,
             )
-            .minus(tokenConversionRate ? ethFee.toString(10) : 0, 10)
+            .minus(tokenConversionRate ? totalEthCost : 0, 10)
```
Comment (danjm) [inline app/scripts/controllers/swaps.js:488]: Review note: this is functionally equivalent. What used to be called `ethFee` is now `totalEthCost`. (What is now `ethFee` is now different from `totalEthCost` when `sourceToken === ETH_SWAPS_TOKEN_ADDRESS)`

File: app/scripts/controllers/swaps.js [hunk 5]
```diff
@@ -394,59 +395,88 @@ export default class SwapsController {
...
+            fromNumericBase: 'BN',
+            numberOfDecimals: 6,
+          },
+        )
+        : totalEthCost
+
       const tokenConversionRate = tokenConversionRates[destinationToken]
       const ethValueOfTrade =
         destinationTokenInfo.symbol === 'ETH'
-          ? calcTokenAmount(destinationAmount, 18).minus(ethFee, 10)
```
Comment (danjm) [inline app/scripts/controllers/swaps.js:449]: Review note: this is functionally equivalent. What used to be called `ethFee` is now `totalEthCost`. (What is now `ethFee` is now different from `totalEthCost` when `sourceToken === ETH_SWAPS_TOKEN_ADDRESS)`

File: app/scripts/controllers/swaps.js (hunk 2 above)
Comment (rekmarks) [reply app/scripts/controllers/swaps.js]: Good suggestion! Done in: 37d1964

**Review by danjm — DISMISSED** [review]
LGTM. I think we might want to hold off on merging until at least until we cut v8.1.2

**Review by brad-decker — APPROVED** [review]
Reviewed live with @rekmarks looks good to me!

**Pull request merged**